### ISSUE TITLE
Accepting Changes for master folder

### ISSUE THREAD

**Issue by nlliqlex** [syncthing/syncthing#2710]
[Feature Request]
Currently Master Folders are really "untouchable". If anything goes out of sync, we have an 'overwrite' button to enforce master's image on all clients.

However, there are times where we wish to accept/reject changes to the master folder for individual items. Could we have an option to be able to do so? This is because I wish to own a master copy of a 'photos' folder, but if my users add in new photos, I wish to be able to accept them individually to update the 'master copy'. 

My rationale for keeping my folder a master copy is because I don't want unauthorized modification of files or accidental deletions on the user's side.

**Comment by calmh** [comment]
An "accept changes" button would probably be fairly simple. However

> accept/reject changes to the master folder for individual items

requires a bit more UI if nothing else.

**Comment by AudriusButkevicius** [comment]
So I disagree with this. I seems that (S)FT is whst you are after, as there is very little syncing left when you have to start having to pick files, approve files and all that crap.

The only possible thing I'd be willing to agree is accept all.

**Comment by lovetox** [comment]
this would be great, i thought about the same thing.

every family member has his own photos on his/her computer.
so they should be able to add photos to the master folder, but not delete them.

this is for all the people who dont want to manage their photos over the google / apple cloud.

@AudriusButkevicius why so negative? syncthing does in this case the same thing as always, it keeps all the folders in sync with the master. the only thing that is changing is that the owner of the master folder can accept additions from other members.

**Comment by AudriusButkevicius** [comment]
I am against it because what you want to achieve can be achieved by an ftp/webdav server with multiple user accounts, no need to wedge that into syncthing.

**Comment by lovetox** [comment]
so either accept all or nothing, everything in between would not be syncthing like?
makes no sense to me, but i guess the "accept all" button would be already an enhancement.

**Comment by ssergiienko** [comment]
+1 For accept changes button (as for me "accept all current changes" is enough)

**Comment by taikobara** [comment]
It would be convenient to be able to automatically accept everything changed on the device, where the master folder is located

**Comment by ProactiveServices** [comment]
> It would be convenient to be able to automatically accept everything changed on the device, where the master folder is located
Surely this is achieved by disabling the Master Folder feature?

**Comment by lovetox** [comment]
> Surely this is achieved by disabling the Master Folder feature?

no, because then every other synced client could make changes aswell

its basically a rights management, it would be nice to have a little bit control over who can add files or delete files from a synced folder.

**Comment by AudriusButkevicius** [comment]
This ticket is not about rights management. It's about a one off accept all changes button which is a shortcut for removing master, waiting to sync, and putting master back on.

**Comment by lovetox** [comment]
yes you are correct, rights management is not really what this is about.

this would be better if i could see what the changes are that i accept, and if this accept prompt would be for a single node, and not many.

for example someone adds something that i want also in the masterfolder, but another node deleted half of his folder cause he doesnt need all files. in this case i wouldnt want to accept all changes.

**Comment by AudriusButkevicius** [comment]
There already exists a feature, called ignoreDeletes.

Als you case about A adding and B deleting is a bit unrealistic as they should be in sync. If they are not in sync, one of them is a master and your changes you choose to accept will most likely be ignored anyway.

**Comment by lovetox** [comment]
I just found this function also in another issue right now

this seems like a good solution for the photo case, no master folder, but a folder which ignores deletes
nodes can add stuff but not delete.

**Comment by taikobara** [comment]
I mean, I would like to be able to only the contents of the master directory was important. Changes on other devices automatically rejected. No matter - the addition or removal. Master the main folder and have the opportunity to reject the changes automatically. It seems to be logical. Another check box - either manually or automatically overwrite changes

**Comment by AudriusButkevicius** [comment]
What you want is not related to this issue, and I bet you can find 5 tickets already closed about this explaning why it makes no sense.

**Comment by taikobara** [comment]
Strange ... In the case where the device from the master folder is remote, and pressing the button in the web interface, it is necessary to specifically login to the remote computer - it may be important. Especially if such devices are many.

**Comment by calmh** [comment]
We should have this in the FAQ. In the meantime, use file permissions to prevent modifications of file on devices where you don't want files to be modified.

**Comment by taikobara** [comment]
What has file permissions?
There are many remote computers. 50.
Information per one new archive is created on a daily basis.
It moved to the Syncthing master folder.
All 50 computers are synchronized with the central (with 50 non-master folder) - it is local, managed computer.
Periodically remote Syncthing detect changes in the central computer folders - actual (sinc-conflict) or false.
And synchronization stops. On the central computer I can not even see that synchronization is stopped. It is necessary to manually search for and go to remote computers to resume synchronization.
I think it would be logical to Syncthing master folder can automatically resolve such conflicts.

**Comment by AudriusButkevicius** [comment]
Well mark all of your files as read only on the filesystem as you sync them to the 50 machines, so that nobody could modify them.

**Comment by taikobara** [comment]
If I mark files as read only - archive rotation stop in the master folder. It is not a way to solve the problem.
Files in this folder are changed, but changes their other program.

**Comment by AudriusButkevicius** [comment]
Regardless, we should stop bikeshedding in this ticket,as it has nothing todo with what you are after.

**Comment by entonio** [comment]
@nlliqlex @lovetox your file-by-file feature request is understandable from a user's perspective, but it's much too specific for a heavy-duty tool such as Syncthing. It might be doable, but it's a bit like wishing for a bookstore to have your favourite books each in a specific font, or for an efficient public transportation system that goes to every street. It's philosophically strained. This is not to dismiss you (and I have nothing to do with syncthing), just to explain why while certainly valuable for some, a file-by-file feature is not likely do be implemented in a tool like syncthing.

BUT ANYWAY: what if you set up separate folders where other uses can write to? Then you can move the files you want into the master folder via drag-and-drop.

**Comment by generalmanager** [comment]
@nlliqlex @lovetox The proper way to setup the photo case is to have different shares with different rights. Everybody has their own subfolder on your devices picture folder. They set their folders to master folder on their devices and have rw rights on your computer. People can write in their folders, but not yours. You don't save shit in their folders and they can't save their stuff in your folders.

There is a reason syncthing takes the simple approach by letting every user create/overwrite files:
Complexity. Once there are more than a few files with a hand full of sub directories, it's not human-managable anymore.

Especially stuff like picture folders is generally very big and sorted badly. There was no private picture folder I've seen with less than 15GB of files, the average was somewhere between 75 and 100 GB. And I've setup syncthing for a LOT of people.

There are also often (sub-)folders with similar or identical names. This means the GUI will have to show the full path from the share, which leads to a lot of clutter.
People also have no idea how software works and just "do their thing". So if they rename a folder, but don't have rights to write on your machine, your computer will have to save the files twice until you manually allow the delete. This is an incredible amount of micromanagement and will always end in a horribly shitty user interface. And everybody trying to g

**Comment by lovetox** [comment]
thanks for telling about your experience, i think thats a good approach with everybody having their own subfolder in a master picture folder. i will try that.

**Comment by generalmanager** [comment]
@lovetox You're welcome. To make sure you get the exact setup I meant above, I'll describe the folder structure and how it's setup in syncthing:

pictures <--this contains all your pictures and is shared with others, but set as a master folder in syncthing
pictures/friends+family <-- this contains all the folders where your friends & family have their stuff and is not synced because it's ignored via the .stignore file
pictures/friends+family/friend1 <-- this is readable and writable by friend1 in syncthing

This achives the following:
* you have everybodies pictures in one folder
* your stuff is never touched by others & vice versa
* you have to setup only one ignore rule and can add more friends later

**Comment by pmatts** [comment]
+1 for "Accept all remote changes" button

Currently I have to
1 - click edit the share
2 - changes master to normal
3 - click save
4 - wait for full synchronisation while executing number 5 (V dangerous!!)
5 - NOT FORGET TO:
6 - click edit
7 - change normal back to master
8 - click save

One button click for this would be very useful to a lot of people and also relatively easy to implement (i imagine)

**Comment by AudriusButkevicius** [comment]
You should reevaluate your setup if you end up doing it often.

**Comment by Ferroin** [comment]
@AudriusButkevicius Had you ever considered that not everybody can use SFTP or set up a WebDAV server?  There's also the fact that most people don't have the luxury of a hosted VPS or some other internet routable system, so they'd need some way to safely relay data between systems if they're not on the same network.  Even if they can use something like that and don't need some relay system, it's generally a bad thing for security (especially if it's regular FTP), and isn't something that you should be doing in general if the system in question isn't intended to be a server, and even then, Syncthing is at least sometimes going to be faster than most alternatives (Syncthing will win out over everything you listed as an alternative if the changes are small and scattered within individual files).

While I understand a desire to help people avoid making poor choices (and I agree that for picking individual changes to merge back, Syncthing is very much the wrong tool, although I'd argue that a proper VCS is more likely what they actually want in that case), this isn't an unreasonable use case, and it should be pretty trivial to implement (add a button to the UI to call a function to let things sync to the master folder until it is fully in sync), the only hard part is figuring out how to handle thing changing on the other side while syncing (and if I had a use case for this myself, the patience to properly learn Go, and sufficient time, I would just write the patch myself)

**Comment by AudriusButkevicius** [comment]
Sure and that is exactly what this ticket is about, one of simple accept all, not sure why you need to empasize it.
I am replying to the last comment specifically, not the thread as a whole, as doing that the author posted repeatedly implies that the folder shouldn't be master to start with, and having to do it once in a blue moon is not so tragic.

**Comment by pmatts** [comment]
This choice makes the software less accessible for two main reasons:

1.  Due to thier lack of understanding, newcomers will naturally be scared of losing thier precious data. (I'm in this category right now). Armed with the knowledge that a master server needs to have a big red button pressed before files are written, they will initially seize their chance of security and set masters.  Before long they will realise they will have to disable their safety net of the big red button in order to incorporate thier remote changes. For a newcomer this is frickin scary! It would be much less scary if they had a big green button making it clear that it will accept thier remote changes as a one off.

2. Logical clarity:  When i was first presented with the big red "override changes" button I immediately asked "override which changes where?" it took some trial and error to discover that what it  was really trying to say was "delete all remote changes".  If the button were to be renamed and another complimentry button "accept all remote changes" added, the new user like me would instantly have a hugely improved understanding of what this software can do simply by looking at two buttons.

You cannot know how people will use your software.

**Comment by AudriusButkevicius** [comment]
Accepting remote changes is scary, as remote changes might be deleting all your files. There is a separate ticket which bikesheds about naming of things, so no need to bikeshed here.

**Comment by pmatts** [comment]
what is bikeshedding? where is this ticket?

**Comment by pmatts** [comment]
i looked up bikeshedding. sorry you feel that way. just trying to help.

**Comment by AudriusButkevicius** [comment]
#2679 is the ticket that talks about various terminologies.

**Comment by ProactiveServices** [comment]
Newcomers to the software should be [consulting the documentation](https://docs.syncthing.net/users/foldermaster.html) if they're scared what might happen to their data.

**Comment by vk496** [comment]
+1 for import all changes.

There are policies for global development or anyone can send pull requests for new features (like this)? 

I'm interested to implement it :)

**Comment by AudriusButkevicius** [comment]
Anyone can send pull requests.

**Comment by marlarius** [comment]
I second this wish very much. While "Override changes" is the obvious choice for a Master folder, there are times where I wish I could just "Accept all changes" this one time instead of temporarily changing the folder type. I do not need a way to accept changes to individual files, though.

**Comment by DavidGP** [comment]
+1 for an **accept changes** (from receive-only folder) **button**,
to prevent accidental mass-deletion in the send-only folder, and still be able to sync your (humble) daily changes back to it...

Thanks heaps for considering!

-- edit; sorry didn't know that +1 comments are not appropriate
-- edit 2 -- OTOH, nobody ever has used an emoticon either on the top of the OP?

**Comment by AudriusButkevicius** [comment]
Please, no +1 comments. Notifies everyone subscribed to the issue/repo sending an email to them and yet adds zero substance to the conversation.

If support some feature use emoticons on the top of the post rather than spamming everyone.
